axlearn/common/attention.py

```diff
@@ -2013,16 +2003,46 @@ def init_states(
                 query and .probs is of shape [batch, num_heads, target_length, source_length].
                 Otherwise, if initializing cache from scratch, output will be None.
         """
-        return self._forward_for_mode(
+        if key is not None:
+            if query.shape[1] != key.shape[1]:
+                raise ValueError("Cross-attention extend_step is not supported.")
+        init_states = dict(time_step=jnp.zeros([query.shape[0]], dtype=jnp.int32))
+
+        # TODO(dhwang2): init_states without prefilling cannot determine whether KV sharing occurs,
+        # so a config option or additional argument should be added.
+        kv_sharing = kv_state is not None
+        if not kv_sharing:
+            kv_shape = KVCache.Shape(
+                batch_size=query.shape[0],
+                kv_len=query.shape[1],
+                num_kv_heads=self.i_proj.num_kv_heads,
+                per_head_dim=self.per_head_dim(),
+            )
+            init_states.update(
+                kv_cache=self.kv_cache.init_states(shape=kv_shape, dtype=query.dtype),
+            )
+
+        if time_step is None:
+            # init_states without prefilling branch.
+            return init_states, None
+
+        # Prefill branch.
+        # TODO(dhwang2): Optimize it by passing only the valid parts of the query. Currently,
+        # prefill has a complexity of O(max_len²), but this can be easily reduced to
```

Comment:
Is this true when we use splash attention?

Reply:
It's general explanation and it's not true on flash attention.